### PR TITLE
Feature/Routing Refactor 1

### PR DESCRIPTION
**Pull request by MrJacobSullivan**
## Description

This PR introduces a `PrivateRoute` component that can provide authenticated, role-based routing. It also refactors unauthenticated routes. It moves all `useEffect` functionality from the `HomeContainer` component. The `PrivateRoute` component will now handle setting the initial redux state as well as getting the profile of the authenticated user from the backend and saving it in state. It utilizes a `useEffect` to handle the `authenticateUser` and the `getProfile` redux functions. I also fixed the way the `NavBar` component is accessing state so that it can handle the case where our `userProfile` object is empty. I also fixed the current front end routing to conform to the route name changes I made.

*Example of the `PrivateRoute` component*

```js
<PrivateRoute
    path="/dashboard"
    component={Dashboard}
    allowRoles={[3, 4]} // limit access to roles using an array of role_id's - this route is limited to mentors and mentees
    redirect="/" // where users will be sent if they do not meet the requirements
/>
```

**Unauthenticated Routes**
`/login` unchanged
`/apply` the `Signup` component, where users can choose between `mentor` or `mentee`
`/apply/mentee` mentee signup form
`apply/mentor` mentor signup form
`/implicit/callback` required by Okta

**Authenticated Routes**
***NONE***

*Routes not implemented in this PR will be implemented in other PR's.*

**Next Steps:**
1. Implement all authenticated routes using `PrivateRoute`
2. Refactor `SideBar` component to `push` to a new route as opposed to rendering a component
3. Handle role-based rending within individual components
    i.e. the same `Dashboard` component will look different for an `admin` and a `mentee`

Trello [#203](https://trello.com/c/57WvaVyh)
[Loom Walkthrough](https://www.loom.com/share/abc15387d2754fdeb2c9110a19ccbd49)
**[Loom Walkthrough Follow-Up](https://www.loom.com/share/0f7ccb4586984dd293afe91bdd819271)**

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have removed unnecessary comments/console logs from my code
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any misspellings
- [x] No duplicate code left within changed files
- [x] Size of pull request kept to a minimum
- [x] Pull request description clearly describes changes made & motivations for said changes

### CODE REVIEW

**Review by Jaredhall18 — APPROVED** [review]
I really like the decision to push the use effect into the private routing.  Your code is very clean and the added notes help clarify specific lines of code.

**Review by Fvaldenegro95 — APPROVED** [review]
Seems like a much more effective implementation! Looks great!

**Review by JohnMBarnett — APPROVED** [review]
Really good idea to go ahead and get private routing implemented. Your video broke down all of your changes nicely, as well as showcased the effect your change will make!

File: src/components/common/PrivateRoute.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  useEffect(() => {
+    if (Object.keys(userProfile).length === 0) {
+      if (profile_id === null) {
+        if (authState.isPending || authState.isAuthenticated) {
+          dispatch(authenticateUser(authState, authService));
+        } else {
+          push(redirect);
+        }
+      } else {
+        dispatch(getProfile(profile_id));
```
Comment (ryansnorek) [inline src/components/common/PrivateRoute.js:33]: Since getProfile is dispatched from within authenticateUser is it still needed here?

**Review by ryansnorek — APPROVED** [review]
This routing refactoring is a big step and looks good so far. Do you know if it would work to use lifecycle.isFetching in redux in place of the loading useState in line 22?

File: src/components/common/PrivateRoute.js (hunk 1 above)
Comment (EnasELM) [reply src/components/common/PrivateRoute.js:33]: awesome, the code is clean also I like how you are explain everything in the video.

**Review by EnasELM — APPROVED** [review]

File: src/components/common/PrivateRoute.js (hunk 1 above)
Comment (MrJacobSullivan) [reply src/components/common/PrivateRoute.js:33]: I moved the `getProfile` function into its own function in PR #[101](https://github.com/BloomTech-Labs/underdog-devs-fe-a/pull/101).

**Review by rdperry031 — APPROVED** [review]
It is extremely satisfying seeing the routing cleaned up. It is good to see private routes being implemented now as well. Looks great.

File: src/components/pages/RoleSignup/Applications/Mentee.js [hunk 2]
```diff
@@ -62,9 +62,9 @@ const Mentee = () => {
           <Breadcrumb.Item href="/login">
             <LoginOutlined />
           </Breadcrumb.Item>
-          <Breadcrumb.Item href="/signup">
+          <Breadcrumb.Item href="/apply">
```
Comment (paulstgermain) [inline src/components/pages/RoleSignup/Applications/Mentee.js:65]: This phrasing for this functionality is so much smoother than 'Signup' - good thinking!

File: src/index.js [hunk 3]
```diff
@@ -67,32 +69,23 @@ function App() {
...
-        <Route path="/pendingapplications" component={PendingApplications} />
-        <Route path="/availability" component={Availability} />
-        <Route path="/schedule" component={Schedule} />
-        <Route path="/manageresources" component={ManageResources} />
-        <Route path="/Profile" component={Profile} />
-        <Route path="/menteesprogress" component={MenteesProgress} />
+        <Route path="/apply" exact component={Signup} />
+        <Route path="/apply/mentee" component={Mentee} />
+        <Route path="/apply/mentor" component={Mentor} />
         <Route path="/implicit/callback" component={LoginCallback} />
```
Comment (paulstgermain) [inline src/index.js:80]: This is looking MUCH cleaner - you love to see it. Great work here!

File: src/components/common/PrivateRoute.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    authState,
+    push,
+  ]);
+
+  return loading ? (
+    <LoadingComponent />
+  ) : (
+    <SecureRoute path={path} component={() => Component({ ...rest })} />
+  );
+};
```
Comment (paulstgermain) [inline src/components/common/PrivateRoute.js:57]: This custom Private Route is brilliant - it does a lot more than a basic Private Route might, and gives out application exactly the kind of security we want!

**Pull request merged**